### ISSUE TITLE
Necessity of the ROOT label when training the dependency parser for custom semantics

### ISSUE THREAD

**Issue by gmastrapas** [explosion/spaCy#2911]
Hi all,

thanks for the awesome work on spaCy! So, I'm trying to train the dependency parser on custom semantics as described [here](https://spacy.io/usage/training#intent-parser). I have some annotated trees with my custom semantics which I transform to a CoNLL file and then to spaCy's JSON training format. I use a slightly modified version of the CLI's training script to train the parser.

My annotated trees look a lot like the ones in the example. Most of the relations are tagged with the '-' tag and only those that matter have tags of interest. The main difference however, is the fact that in my data, the root of the tree can have any relation label from the set of labels and not just the 'ROOT' label. Think of this like the word 'find' being root and having a 'PLACE' label in the example.

I ran some training sessions and got some poor results in my validation set. I quickly found out that my trained parsers always tag the root of the tree with the 'ROOT' label, even though the specific label is not present in my data. Let me add that my annotated trees can be non projective so I was using projectivize=True when generating train_docs from my corpus. I found out that the argument was invoking a preprocess_training_data function that was changing all the root labels in my training data to the 'ROOT' tag. I changed this and recompiled but still the problem persists. The only difference is that now I get poor results in the training set as well which is reasonable.

Can I actually do this by slightly modifying the source code? Or is there a necessity in the underlying algorithm for the root label to be 'ROOT' (The code inside arc_eager.pyx seems to point that way).

I thought about including a dummy <root> word in all my data that has always the 'ROOT' tag and have my actual root or roots have this <root> word as head with the actual root labels. I am not sure this makes sense though.

Any help would be much appreciated! Thanks!  

## Your Environment

* Operating System: OSX 10.14
* Python Version Used: Python 3.6.4
* spaCy Version Used: spaCy 2.0.12
* Environment Information: -

**Comment by honnibal** [comment]
Unfortunately you'll have to run a separate model to tag your root words. It's ugly but the problem is, the transition system halts when there's only one word on the stack. It doesn't make a prediction over this word, although that would be a reasonable thing to do.

The good news is, you'll quite likely get better results by running another model.
